crypto/pkcs12/p12_crt.c

```diff
@@ -69,12 +70,18 @@ PKCS12 *PKCS12_create_ex(const char *pass, const char *name, EVP_PKEY *pkey,
             goto err;
         if (keyidlen && !PKCS12_add_localkeyid(bag, keyid, keyidlen))
             goto err;
+        if (cb)
+            if (cb(bag, cbarg) == -1)
+                ERR_raise(ERR_LIB_PKCS12, PKCS12_R_CALLBACK_FAILED);
     }
 
     /* Add all other certificates */
     for (i = 0; i < sk_X509_num(ca); i++) {
-        if (!PKCS12_add_cert(&bags, sk_X509_value(ca, i)))
+        if ((bag = PKCS12_add_cert(&bags, sk_X509_value(ca, i))) == NULL)
             goto err;
+        if (cb)
+            if (cb(bag, cbarg) == -1)
```

Comment:
So we haven't implemented my suggestion of a 0 return meaning don't add the bag after all. I still think that might be useful. But, if we don't do that, then probably the return values should be 1 for success and 0 for error.

Reply:
@mattcaswell I am confused now. Didn't you mean to add the callback that adds optional attributes to the bag rather than the callback to add the bag to PKCS12?

Or do we now have two callbacks?

Reply:
Yeah the call back is to add optional attrs to the safe bag

Reply:
So if the `friendlyName` and `localKeyId` are added to the bag before the callback via some other mechanism, then we should not discard the whole bag or even the existing attributes if adding optional attributes fails?

I agree that the callback should return success / failure, but I am not sure whether the new API code calling the callback should do anything with the callback return value other than stopping calling the callback on failure and returning.

Reply:
I see there being one callback. Its purpose is to do optional application processing on the bag. One such use is to add attributes to the bag (which is what this particular PR is going to be doing). It could be used for anything, e.g. to add friendly names to all the "other" certs if the application so wishes (not sure why you would). This PR is only using the callback to add optional attributes, but the public API allows an application to do whatever it likes.

Another use could be to do filtering on the certs that get exported. So the `PKCS12_create_ex2()` call could supply a large number of certificates and the callback could be used to do some filtering and determine only a subset to be actually exported (perhaps based on some local policy rules). 

My suggestion is that in the event that the return value from cb is 0, then remove `bag` from `bags`. This would also remove any friendly name/keyid added to the bag...but that's ok.

Reply:
@mattcaswell I am not sure I like that "multi-purpose callback" idea.

Also, what happens if the passed in bag is say a shrouded keybag, unless that's impossible?

If it is possible, does the callback even have enough info to process it (i.e. can it see inside or can it just see the attributes and what can it do if there aren't any)?

Reply:
> I am not sure I like that "multi-purpose callback" idea.

Whether we like it or not, the callback as currently in this PR *is* multi purpose. We cannot restrict what an application might do in that callback. Nor do I think we want to. The idea behind the callback is to make pkcs12 creation more flexible.

Reply:
is there an existing API to remove a `bag` from `bags`? or should I add one in pk12_crt?